cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs

```diff
@@ -0,0 +1,353 @@
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go, env: " ++ show env
+    shutdown <- gets (\RWEnv{terminating, inFlight, notYetSent} ->
+                       terminating &&
+                       Seq.null inFlight &&
+                       Seq.null notYetSent)
+    if shutdown
+    then do -- terminating normally, inform local peer
+      -- check if remote peer is waiting on us
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "TERMINATING normally"
+      (opP, op) <- maybe (False, error "internal error")
+                         (\x -> (True,snd x)) <$> gets availableOp
+--        liftIO $ putStrLn $ "TERMINATING normally, opP: " ++ show opP
+      lift . atomically $
+        if opP
+        then putTMVar op Nothing
+        else do
+          reqTxIds <- takeTMVar rpcIn -- no other message should occur as nothing is in flight
+          case reqTxIds of
+            RPCRequestTxIds _ resp -> putTMVar resp Nothing
+            _ -> return ()
+    else do -- process next interaction
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "go, process next interaction"
+      go1 env
+
+  go1 :: ROEnv txid tx -> StateT (RWEnv m txid tx) m ()
+  go1 env = do
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, env: " ++ show env
+    terminating' <- gets terminating
+    notYetSent'  <- gets notYetSent
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, terminating': " ++ show terminating'
+
+    -- perform the blocking inquiry
+    (nowTerminating, newTxs, rpc) <- lift . atomically $ do
+      -- just looking for the transition to terminating here
+      term' <- if terminating'
+               then pure False
+               else readTVar termVar
+      -- try get more txs if our target backlog permits
+      txs' <- if Seq.length notYetSent' < targetBacklog env &&
+                 not terminating'
+              then tryTakeTMVar txIn
+              else pure Nothing
+      -- always willing to interact with local peer
+      rpc' <- tryTakeTMVar rpcIn
+      -- check there was some change to process
+      unless (term' || isJust txs' || isJust rpc') retry
+      -- treat empty input transaction list as equiv to
+      -- terminating
+      let (term,txs) =
+            case txs' of
+              Just x | null x
+                       -> (True, Nothing)
+              _        -> (term', txs')
+      pure (term || terminating', txs, rpc')
+
+    -- Update terminating, if needed 
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, update terminating, if needed"
+    when (nowTerminating && not terminating') $ do
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, update terminating: True"
+      modify (\x -> x {terminating = True})
+
+    -- incorporate new transactions from the generator into
+    -- NotYetSent
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, incorporate new transactions from the generator into NotYetSent"
+    flip (maybe (pure ())) newTxs $ \xs -> do
```

Comment:
Hmm.. is it just me, or is this somewhat hard to read? 

Do you think we could have a simpler presentation -- maybe a dedicated function?

Reply:
This code was written by Neil. To be honest, I don't want to touch it right now (actually, only Neil really knows how this code works).